src/java/org/apache/ivy/util/url/HttpClientHandler.java

```diff
@@ -322,13 +327,34 @@ private CloseableHttpResponse doHead(final URL url, final int connectionTimeout,
                 .build();
         final HttpHead httpHead = new HttpHead(normalizeToString(url));
         httpHead.setConfig(requestConfig);
-        return this.httpClient.execute(httpHead);
+        return this.httpClient.execute(httpHead, getHttpClientContext(url));
     }
 
     private boolean hasCredentialsConfigured(final URL url) {
         return CredentialsStore.INSTANCE.hasCredentials(url.getHost());
     }
 
+    private HttpClientContext getHttpClientContext(URL dest) {
+        if (!getPreferPreemptiveAuth()) {
+            return null;
+        }
+
+        // Preemptive authentication, see
+        // https://hc.apache.org/httpcomponents-client-4.5.x/tutorial/html/authentication.html
+        // Without preemptive authentication, Ivy will first try non-authenticated, get a 4xx, retry
+        // authenticated and finally get a 2xx. This becomes an issue in an environment where Ivy is
+        // used a lot, as firewalls might see large amounts of 4xx as a brute-force attack and close
+        // the connections.
+
+        final BasicScheme basicAuth = new BasicScheme();
```

Comment:
What if there is a difference authentication scheme used? The order of preference is namely digest, basic, ntml, so basic is not the first choice.

Reply:
Good catch! I've just checked HttpClient's documentation, and I believe I'd need to be able to attach 3 `AuthCache`s to the `HttpClientContext` via some kind of `AuthSchemeRegistry` or `AuthSchemeProvider`, unfortunately there is no such thing.

Also there is only one implementation of `AuthCache` anyway: `BasicAuthCache` (there is no such thing as `DigestAuthCache` or `NTLMAuthCache`). It's normal I believe, and apparently preemptive auth makes only sense with basic auth, because for Digest and NTLM you need to receive a salt/challenge/key first in order to be able to send your creds accordingly, so the creds are not static things that you can send upfront.

Nevertheless you're right: I have no clue how this behaves if the authentication scheme isn't Basic and I'll have to figure this out before trying to get this PR upstream.